### PR TITLE
Fix duplicate "Preserved directory X" messages when running tests with PRESERVE=1

### PR DESCRIPTION
**Pull request by grossag**
This change fixes duplicate "Preserved directory X" messages when running tests with PRESERVE=1 in the environment. The root cause was that self._dirlist wasn't being cleared in this codepath.

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [ ] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: src/CHANGES.txt [hunk 1]
```diff
@@ -6,6 +6,11 @@
 
 RELEASE  VERSION/DATE TO BE FILLED IN LATER
 
+  From Adam Gross:
+
+    - Fixed duplicate "Preserved directory X" messages when running unit tests
+      with PRESERVE=1 in the environment.
```
Comment (bdbaddog) [inline src/CHANGES.txt]: Please change "unit tests" to "SCons unit tests" just to avoid any confusion that it might impact other non-scons tests which might be run by SCons in a users build.

**Comment by grossag** [comment]
> Will need a fix to the test-framework unit test (which is not automatically run in CI) as well: `testing/framework/TestCmdTests.py`
> 
> ```
> ======================================================================
> ERROR: test_preserve (__main__.preserve_TestCase)
> Test preserve()
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "testing/framework/TestCmdTests.py", line 1507, in test_preserve
>     assert not os.path.exists(wdir)
> AssertionError
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "testing/framework/TestCmdTests.py", line 1511, in test_preserve
>     test._dirlist.remove(wdir)
> ValueError: list.remove(x): x not in list
> ```

Thanks for the pointer. TestCmdTests.py assumes that _dirlist is preserved across calls to cleanup() and that multiple calls to cleanup() are expected. I am going to remove that assumption from the tests; let me know if you have a problem with that.

**Comment by grossag** [comment]
I updated the tests. I am a bit confused now in that the existing behavior seems to have been desired, although it led to confusing behavior. Please let me know if we should just drop this pull request instead of making these changes.

**Comment by mwichmann** [comment]
Personally, my experience is there's a lot of "test made to match observed behavior" rather than the other way around (test written to express desired behavior, then code implemented can be shown to correctly follow that behavior).  We'll have to leave the response to one of the maintainers...

File: testing/framework/TestCmdTests.py [hunk 2]
```diff
@@ -1503,12 +1503,11 @@ def cleanup_test(test, cond=None, stdout=""):
             test.preserve('pass_test')
             cleanup_test(test, 'pass_test', "Preserved directory %s\n" % wdir)
             assert os.path.isdir(wdir)
-            cleanup_test(test, 'fail_test')
+            shutil.rmtree(wdir, ignore_errors = 1)
```
Comment (bdbaddog) [inline testing/framework/TestCmdTests.py:1506]: Why this change?

Comment (grossag) [reply testing/framework/TestCmdTests.py:1506]: The tests were relying on behavior where cleanup_test can be called many times with a 'pass_test' result. Then one you call it with a 'fail_test' result, it will remove all directories. AFAICT this is not normal behavior; a test will either pass or succeed and we should clean up once.

Comment (bdbaddog) [reply testing/framework/TestCmdTests.py:1506]: The issue with this change is that it's not respecting the PRESERVE, PRESERVE_FAIL shell environment variables not deleting test directories when set..

**Comment by bdbaddog** [comment]
I think this is not the best solution.
Seems like the issue is multiple of the same directory getting added to test._dirlist?
If so wouldn't it make more sense to avoid adding duplicates?

Maybe here: https://github.com/SCons/scons/blob/master/testing/framework/TestCmd.py#L1760

**Comment by grossag** [comment]
I don’t necessarily think it’s worth spending a bunch of time on this PR. Maybe just best to close it.

**Comment by bdbaddog** [comment]
probably not. It's an annoyance. If you point me at a test which does this I'll see if I can resolve.
Closing.
